source/crypto.c

```diff
@@ -238,3 +241,236 @@ PyObject *aws_py_hmac_digest(PyObject *self, PyObject *args) {
+    struct aws_allocator *allocator = aws_py_get_allocator();
+    struct aws_array_list pem_list;
+    if (aws_pem_objects_init_from_file_contents(&pem_list, allocator, pem_data_cur)) {
+        return PyErr_AwsLastError();
+    }
+
+    struct aws_pem_object *found_pem_object = s_find_pem_object(&pem_list, AWS_PEM_TYPE_PRIVATE_RSA_PKCS1);
+
+    if (found_pem_object == NULL) {
+        PyErr_SetString(PyExc_ValueError, "RSA private key not found in PEM.");
+        goto on_done;
+    }
+
+    struct aws_rsa_key_pair *key_pair =
+        aws_rsa_key_pair_new_from_private_key_pkcs1(allocator, aws_byte_cursor_from_buf(&found_pem_object->data));
+
+    if (key_pair == NULL) {
+        PyErr_AwsLastError();
+        goto on_done;
+    }
+
+    capsule = PyCapsule_New(key_pair, s_capsule_name_rsa, s_rsa_destructor);
+
+    if (capsule == NULL) {
+        aws_rsa_key_pair_release(key_pair);
+    }
+
+on_done:
+    aws_pem_objects_clean_up(&pem_list);
+    return capsule;
+}
+
+PyObject *aws_py_rsa_public_key_from_pem_data(PyObject *self, PyObject *args) {
+    (void)self;
+
+    const char *pem_data_ptr;
+    Py_ssize_t pem_data_len;
+    if (!PyArg_ParseTuple(args, "s#", &pem_data_ptr, &pem_data_len)) {
+        return NULL;
+    }
+
+    PyObject *capsule = NULL;
+    struct aws_byte_cursor pem_data_cur = aws_byte_cursor_from_array(pem_data_ptr, pem_data_len);
+    struct aws_allocator *allocator = aws_py_get_allocator();
+    struct aws_array_list pem_list;
+    if (aws_pem_objects_init_from_file_contents(&pem_list, allocator, pem_data_cur)) {
+        return PyErr_AwsLastError();
+    }
+
+    struct aws_pem_object *found_pem_object = s_find_pem_object(&pem_list, AWS_PEM_TYPE_PUBLIC_RSA_PKCS1);
+
+    if (found_pem_object == NULL) {
+        PyErr_SetString(PyExc_ValueError, "RSA public key not found in PEM.");
+        goto on_done;
+    }
+
+    struct aws_rsa_key_pair *key_pair =
+        aws_rsa_key_pair_new_from_public_key_pkcs1(allocator, aws_byte_cursor_from_buf(&found_pem_object->data));
+
+    if (key_pair == NULL) {
+        PyErr_AwsLastError();
+        goto on_done;
+    }
+
+    capsule = PyCapsule_New(key_pair, s_capsule_name_rsa, s_rsa_destructor);
+
+    if (capsule == NULL) {
+        aws_rsa_key_pair_release(key_pair);
+    }
+
+on_done:
+    aws_pem_objects_clean_up(&pem_list);
+    return capsule;
+}
+
+PyObject *aws_py_rsa_encrypt(PyObject *self, PyObject *args) {
+    (void)self;
+
+    struct aws_allocator *allocator = aws_py_get_allocator();
+    PyObject *rsa_capsule = NULL;
+    int encrypt_algo = 0;
+    const char *plaintext_ptr;
+    Py_ssize_t plaintext_len;
+    if (!PyArg_ParseTuple(args, "Ois#", &rsa_capsule, &encrypt_algo, &plaintext_ptr, &plaintext_len)) {
+        return NULL;
+    }
+
+    struct aws_rsa_key_pair *rsa = PyCapsule_GetPointer(rsa_capsule, s_capsule_name_rsa);
+    if (rsa == NULL) {
+        return PyErr_AwsLastError();
```

Comment:
PyCapsule_GetPointer() already sets a good python exception if it fails. Calling PyErr_AwsLastError() will overwrite it with nonsense from whatever stale value aws_last_error() returns.
```suggestion
    struct aws_rsa_key_pair *rsa = PyCapsule_GetPointer(rsa_capsule, s_capsule_name_rsa);
    if (rsa == NULL) {
        return NULL;
```

Reply:
ah, was copying the behavior from sha impl earlier in the file. will fix